aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsAsyncContext.java

```diff
@@ -68,16 +68,15 @@ public boolean hasOriginalRequestAndResponse() {
 
     @Override
     public void dispatch() {
-        try {
-            log.debug("Dispatching request");
-            if (dispatched.get()) {
-                throw new IllegalStateException("Dispatching already started");
-            }
+        log.debug("Dispatching request");
+        if (dispatched.get()) {
+            throw new IllegalStateException("Dispatching already started");
+        }
+        if (!dispatchStarted.get()) {
+            dispatchStarted.set(true);
+        } else {
             dispatched.set(true);
-            handler.doFilter(req, res, ((AwsServletContext)req.getServletContext()).getServletForPath(req.getRequestURI()));
             notifyListeners(NotificationType.START_ASYNC, null);
-        } catch (ServletException | IOException e) {
-            notifyListeners(NotificationType.ERROR, e);
         }
     }
```

Comment:
We could improve this a little bit, using functionality specific to AtomicBoolean, to eventually prevent a race condition (although I'm not sure how important it is here). 

But this is equivalent to:

```
       if (dispatchStarted.getAndSet(true)) {
           dispatched.set(true);
           notifyListeners(NotificationType.START_ASYNC, null);
       }
```

You probably looked at this code more times than me, but let me know if this makes sense to you.

Reply:
It seems better to change the code as you suggested, for the sake of atomicity.